loki/bulk/item.py

```diff
@@ -628,3 +653,141 @@ def calls(self):
             return (type_.initial.name.lower(), )
         # The name of the procedure is identical to the name of the binding
         return (name_parts[1],)
+
+
+class GenericImportItem(Item):
+    """
+    Implementation of :class:`Item` to represent a generic Fortran module import.
+
+    This does not constitute a work item when applying transformations across the
+    call tree in the :any:`Scheduler` and is skipped during the processing phase.
+    However, it is needed to determine whether the import corresponds to a type
+    declaration or a procedure interface.
+    """
+
+    def __init__(self, name, source, config=None):
+        name_parts = name.split('#')
+        assert len(name_parts) > 1 #only accept fully-qualified module imports
+        super().__init__(name, source, config)
+        assert self.scope
+
+    @property
+    def routine(self):
+        """
+        Always returns `None` as this is not associated with a :any:`Subroutine`
+        """
+        return None
+
+    @property
+    def members(self):
+        """
+        Empty tuple as generic imports have no member routines
+
+        Returns
+        -------
+        tuple
+        """
+        return ()
+
+    @property
+    def function_interfaces(self):
+        """
+        Empty tuple as generic import items cannot include interface blocks
+
+        Returns
+        -------
+        tuple
+        """
+        return ()
+
+    @cached_property
+    def imports(self):
+        """
+        Return modules imported in the parent scope
+        """
+        return self.scope.imports
+
+    @property
+    def calls(self):
+        return ()
+
+    @property
+    def procedure_interface_members(self):
+        """
+        The set of children unique to items of type :class:`GenericImportItem`.
+        Comprises exclusively of function calls bound to a procedure interface.
+        """
+
+        _children = ()
+        intfs = self.scope.interfaces
+        if isinstance(self.source[self.local_name], TypeDef):
+            pass
+        elif self.local_name in [i.spec.name for i in intfs]:
+            for i in intfs:
+                if i.spec.name == self.local_name:
+                    for b in i.body:
+                        if isinstance(b, ProcedureDeclaration):
+                            for s in b.symbols:
+                                _children += as_tuple(s.name)
+                        elif isinstance(b, Subroutine):
+                            if b.is_function:
+                                _children += as_tuple(b.name)
+
+        return _children
+
+
+class GlobalVarImportItem(Item):
+    """
+    Implementation of :class:`Item` to represent a global variable import.
+    These only appear at the end of dependency trees and do not have children.
```

Comment:
Considering the ball of wool that our dependency trees (dependency shrubbery?) tend to resemble, I think "end of dependency tree" may not be well defined. The intention is that it doesn't have any dependencies itself, right?

What about (not uncommon) situations like this:

```f90
module some_mod
  use parkind1, only: jprb

  real(kind=jprb) :: var
end module some_mod
```

with

```f90
subroutine some_routine
  use some_mod, only: var
  ...
end subroutine some_routine
```

Should this generate another import dependency? I would (at least in principal) expect something like
```
SubroutineItem(#some_routine) --> GlobalVarImportItem(some_mod#var) --> GlobalVarImportItem(parkind1#jprb)
```
although that is probably tricky to derive in practice (and likely not necessary for our purposes).

 Maybe worth adding a test for that scenario to clarify the behaviour?

Reply:
Firstly, dependency shrubbery is spot on 😆 what I wanted to say here is that a `GlobalVarImportItem` does not have any children (I intentionally left out the mechanism to add children). So in your example, `some_mod#var` does indeed depend on `parkind#jprb`. But I intended `GlobalVarImportItem` to just be dedicated to variables that are used to store data, and not a catchall for any import.

According to my (probably quite wooly to continue the analogy) nomenclature, `parkind1#jprb` would be a `GenericImportItem`. But I didn't add a mechanism to add `GenericImportItem`s as children of `GlobalVarImportItem` because I didn't think there would be any practical need for it.

I can certainly add a test to clarify this behaviour. But do let me know please if you would like me to update the naming convention to something clearer, or even change the behaviour of `GlobalVarImportItem` to have `GenericImportItem` children, even if we don't do anything with them.